### DISCUSSION TITLE
ZIP Framework Ideas

### DISCUSSION THREAD

**Discussion by georgezgeorgez** [Big-Inches-Club-House/bich#4] [BICH Proposals]
Note: I haven’t done any research on how other networks do it

- The network = Distributed nodes, decentralized operators, running software which adheres to common protocol
- At its purest the ZIP framework is to provide structured facilitation from idea to software upgrade.
- Distributed and decentralized aspects of the network present coordination challenges
- We define a set of opt-in software delivery lifecycle practices

High level practices:
1) Sharing of ideas 
2) Feedback and discussion 
3) Signaling support or opposition 
4) Software Implementation
5) Testing Requirements 
6) Acceptance and Integration
7) Software distribution
8) Upgrade/Activation

For all these activities, need to balance decentralization, and efficiency/convenience
- In particular the community is not sure what level of involvement the founding dev team is willing or wants to have
- My recommended approach is to assume minimal involvement where possible while also allowing opt-in opportunities for founding devs to participate
- If any centralized actors are needed, remove any subjective decision making as much as possible and have the actor simply be an executor

But first need to consider the scope of the ZIP framework:
- Protocol?
- Implementations?
- Clients?
- Contract standards?
- Off-chain ecosystem and integrations?
- Official website and marketing?
- Etc

For things not directly related to the protocol or node software, hard to say what should or shouldn’t be in scope of the zip framework
When thinking about decentralization, maybe we don’t need to decide though

Important to make governance distinction between protocol and software, definition vs implementation 
- As a matter of practicality, we start with a single implementation that implicitly defines the protocol
- Long term, need to abstract the protocol away from the reference implementation 

Benefits of multiple implementations: 
1) protocol is more intimately accessible to a wider dev community
2) implementation specific bug which halts the node won’t halt the network

Drawbacks of multiple implementations:
1) More work needed to roll out protocol changes 
2) Non halting protocol related bugs could split the network
What’s the right balance?
- Ideally network is fault tolerant to maintain supermajority validator/nodes uptime in case of halting bug
- 3-4 implementations with uniform adoption? Or maybe 2 is enough with validators/nodes running both in an HA setup?
- Pure implementation changes are much smaller in scope than protocol changes

Breaking down each step of the zip framework

1) Sharing of ideas
- Primary concerns: Naming and storage of documents
- Publishing on the web is easy but curated stable storage with easy to use identifiers is efficient
- The documents don’t really need to follow a standard structure or provide a structured metadata unless we are writing tooling to use them
- People are used to single central repository with a few editors/curators; this is easy to find and navigate
- But this places a subjective burden on editors on what to include and also creates a single point that can very easily be disrupted
- With recent events, it seems many other projects are questioning their reliance on central repositories and maintainers 
- Long term, ideally we use services built on top the network, but we’re not there yet
- In theory my decentralized ideal would be immutable documents hosted via something like bittorrent or ipfs, where anyone can collectively support the storage/retrieval of things they want to share. Anyone would be able to curate/compile lists for others to reference since they all point to the same things
- These documents are pretty small in size and such granularity of data replication is likely not needed
- Git repositories are the expected de facto standard interface and provide decentralized data replication mechanisms
- Git repositories support submodules to include other repositories
- Our network governance is done through Pillars as representatives. 
- We can draw from real world representative governance; proposed legislation must be sponsored/introduced by a representative
- This is a logical requirement because any proposals that aren’t able to find a sponsoring representative are either Dead on Arrival, or worse: voted on without anyone’s understanding. Sponsorship is reputation skin in the game
- My suggestion is thus that all ZIPs must be formally introduced via a sponsoring Pillar. Sponsorship implies that the pillar will vote for it, and ultimately upgrade software if needed. A change that cannot find a sponsor is either underdeveloped or dead on arrival.
- Changes that no pillar cares about are also arguably out of scope of the ZIP framework 
- Pillars can introduce/sponsor ZIPs by hosting it in their own git repo. They can verify ownership by including a signature in the repo. Ideally they follow a sequential numbering scheme with file naming
- This creates a convenient and unique identification scheme: ZIP \<sponsor name\>: \<file path\>
- Supporters can mirror the repo for data redundancy, ideally on other platforms
- Aggregation can be done by anyone through pointer repos that include zip repos through git submodules
- The role of optional official editors: maintain an official pointer repo, verify pillar ownership of their zip repos (through signature file), and add as submodule. The only subjective decision making needed is for abuse from pillars, eg obscene content 

To be continued for other steps

**Comment by 0x3639** [comment]
"Aggregation can be done by anyone through pointer repos that include zip repos through git submodules". 
Q: Does this mean the central list of ZIPs would be pointers to lots of other repos?  Can github point to a gitlab repo?  Can pillars also submit a PR to the list repo with a pointer to their ZIP?  

So either the central maintainer can add them.  Or the Pillar can add a PR to add them.

**Comment by 0x3639** [comment]
"In theory my decentralized ideal would be immutable documents hosted via something like bittorrent of ipfs, where anyone can collectively support the storage/retrieval of things they want to share. Anyone would be able to curate/compile lists for others to reference since they all point to the same things". 

I was thinking about spinning up my own gitlab and host it myself to implement your ideas.  That is not backed by IPFS, but I control it.  It cannot be easily taken down.  I can then sync to github for backup.  How are you thinking that Pillars host ZIPs on IPFS or Bittorrent.  That would make pointing to them on a central ZIP list / number repo difficult??  Maybe for those unable to spin up Gitlab posting a document to IPFS could work.  Is that what you were thinking?

**Reply by georgezgeorgez** [reply]
Oops I meant bittorrent OR ipfs not "of"
So that example was to give what my ideal would be. Maybe long term. Decentralized hosting of immutable documents where anyone who supports the document can contribute to storing it and making it available for download (seeding/pinning).

But most people don't have experience or aren't willing to run a torrent tracker, ipfs gateway, or simply seed/pin.

Git is the de facto standard so for now we can use that as the building block. A repo is easy to create and anyone can still support its availability by mirroring it.

https://git-scm.com/book/en/v2/Git-Tools-Submodules

Git submodules creates an explicit reference to another repo.

But actually, there's no reason why a link in the README wouldn't work as well. Submodules could be over complicating it.

Basically the core idea is, our network has representatives with non-trivial stake in the network, so let's put the balance of centralization/decentralization at that level.

**Reply by 0x3639** [reply]
Just so I'm 100% clear.  As a Pillar I could accomplish this for deeZNNutz by spinning up gitlab on my infra.  And host my own ZIPs for the community.  

I create a repo called ZIPs.  I then create my own process for community members to submit a ZIP for me to review.  And if I support it I publish it as ZIP:deeZNNutz:0001 on my gitlab instance.  We can sync that to github for backup.  and the community can replicate / backup to.  

By publishing a ZIP on deeZNNuts it means I support it and will communicate to other pillar that I support it and encourage them to support it too.  

This will eventually lead to code being hosted on various repositories run by various Pillars.  Over time pillars will chose to adopt code on my repo or not.  Node software could be forked and supported by 50% of pillars.  The other 50% could support a different node software.  But the consensus is the same  

Am I understanding that correctly?

**Reply by georgezgeorgez** [reply]
I've only made suggestions so far regarding 1) sharing of ideas. Which may not have accompanying code changes yet. But a lot of the same principles can apply to the later steps for Software Distribution etc. But you're absolutely right that anyone can run whatever software they want as long as it still conforms to the rules of the protocol.

There are layers of possible censorship.
- repo
- service/platform (eg GitHub, Gitlab SaaS)
- hosting (eg aws, your own metal)

The deeper you go with decentralization, the smaller the blast radius of censorship.
 
So yup you could run a git server on your own infra/metal (For most people, I would suggest gitea over gitlab. Gitlab is great but has a lot of moving parts. I've done enterprise deployments of gitlab before ahah)

Others might just create repos on hosted GitHub or Gitlab. Or maybe even on your git server. 

As you said, you could have your own ZIP repo and process for proposals that you will sponsor. I think one of the implicit themes of our network is connecting/aligning people/validators who have the same values.

The great thing is that it's very easy to mirror repositories.

https://docs.gitea.io/en-us/repo-mirror/

https://docs.gitlab.com/ee/user/project/repository/mirror/

So in additional to critical software components, it would be easy for you to backup other pillar's zip repos that you support.

If we have a few community members running mirrors on separate infra, we're pretty resilient. 

The issue now is that we have a bunch of separate pillar zip repos but no way for someone to explore them all easily.

Which is where the pointer repos comes in. They reference the pillar specific repos. You can just use a link in the readme, but the benefit of git submodules is that you can recursively clone.

If you're running your own git server mirroring the repos of other pillar zips, you can create a pointer repo that only points to the mirrors that you control.

The official Zenon GitHub org can do the same thing as you. They don't maintain a pillar specific zip repo but can maintain a pointer repo, and mirrors of things off GitHub if they wanted.

**Reply by 0x3639** [reply]
So no "centralized" repo pointer?  It's not necessary if everyone is pointing to each other I suppose.  But could make it easy is someone just took on this task of setting up a central repo that just points to all the other ZIP repos.  AND, the Pillars could do the same for the ZIPS they like / support.  

.... but to be clear, we could have a github "ZNS" repo that simply points to ALL the ZIPs supported by the pillars.  And those pointers will be all over the place.... github, gitlab, ipfs, where ever!!!

**Reply by georgezgeorgez** [reply]
Yes exactly.

Anyone could have their own pointer repo which references pillar zip repos.

A centralized one in the official GitHub is not needed but can help a lot for newcomers. And there is not too much subjective decision making needed to maintain it. It also shouldn't be that much work considering the rate of pillar growth.

Many pillars likely won't sponsor zips themselves. Just as in irl representative politics, some representatives end up leading.

This isn't an issue because other representatives are not restricted. Of course you could have multiple sponsors for an initiative.

**Comment by 0x3639** [comment]
Working with this idea, I created the following:

https://github.com/orgs/deeZNNutz-com/repositories

Folder for Proof of Pillar
Folder for ZIPs

Added discussion board for community to submit ZIPs for consideration.  Setup gitea in offsite datacenter for replication and redundancy.  Open to feedback on setup.

**Comment by 0x3639** [comment]
ZIP Pointer Service = Main Repo on Github that points to other Pillar Repos (Submodules)
1st Pillar Repo = dZIP.deeZNNutz.com
in the repo dZIP.deeZNNutz.com we will house markdown files for dZIPs proposed by the community and supported by deeZNNutz.com
dZIP.deeZNNutz.com, ZIP Pointer Service, zenon-network and others will get mirrored to gitea offsite - one way push from Github to gitea

I feel like we need a good naming convention.  I think what I proposed sucks.  Thoughts?

repo name = dZIP.[NAME OF PILLAR]
dZIPs are numbered 0001.md in the repo above.

**Comment by 0x3639** [comment]
[dZIP.pdf](https://github.com/Big-Inches-Club-House/bich/files/9434392/dZIP.pdf)

**Comment by georgezgeorgez** [comment]
My goal in this thread is to share ideas about what a ZIP framework could look like.

I invite anyone to contribute and discuss. 

As I've said in previous messages, the framework is ultimately an opt-in process

I've identified a few high level steps and began by sharing some ideas regarding how documents can be stored and identified. 

For each step/concern, we must strike a balance between practicality and decentralization. 

It's likely that over time, as our community and technology matures, our framework will as well.

We also have no way to compel any participants in our community. Such is the nature of decentralized networks. We can only align incentives.

There has been some existing work done towards defining a ZIP framework. Unfortunately I do not think those current recommendations fully consider what an end-to-end Idea to Decentralized Network Upgrade entails.

In addition, it suggests that communities members can compel other members, which we find incompatible with our values.

Generally I do not like to focus my efforts on criticizing the work of others. (Instead their principles and attitudes). But I do want to make note that there is an AZ funding proposal attached, which I believe entails a level of scrutiny.

As part of the pre-existing work, a vote is being conducted via AZ to approve its guidelines.

I do not find AZ as is to be a good polling mechanism because of its time restrictions and quorum behavior. For example, something will pass AZ if only 1/3 of pillars vote and more than half of those votes are yes. With a minimum quorum, it is possible to pass AZ with only just more than 1/6 of pillars voting yes. 

Generally for community consensus, we would like to have at the minimum a supermajority 67%. In practice, an activation requirement is often set to 90%.

The proposal itself has been accepted at under 30 pillars reporting and the phase to close it seems to be still be open. (Btw it is not needed to close out a project. It also hurts the case that the initial proposal vote has meaning.)

In response to the ideas being discussed in this thread, it is being asserted and heavily emphasized that the existing guidelines have already been "accepted". It is strongly implied that our work and ideas must be compatible with the existing work to have any merit.

From a de-jure perspective, I do not believe a strong consensus has been met. 

From a de-facto perspective, a vote has no meaning in its own. It must be enforced. And in our network, we have no way to compel or enforce off-chain behavior from anyone. Only opt-in processes

From the perspective of my efforts to help share ideas, the assertion of "acceptance" has little impact.

But from the perspective of  community standards and what decentralization means, I find it necessary to address it further.

I have thus created an AZ proposal such as the one for the existing work. It is titled "BICH ZIP ideas"

Vote NO - if there is no need to further discuss in this thread the ZIP process because it has already been accepted and that you will adhere or try to enforce it.

Vote YES - if any "acceptance" of the existing proposal is non binding and that you want to see ideas regarding the rest of the steps, even if they do not conform to the existing work.

Ultimately, I think it is likely that the way we settle on a de facto framework is by actually trying to push a piece of work through and seeing what works and what the community, pillars, builders, users, will opt-in to. 

It is not strictly necessary to start with a ZIP to define the ZIP process. Rather it can first be worked out in practice and then formalized later.

**Reply by 0x3639** [reply]
Well said. I want to keep talking about this so we get it right.  I also believe trying a ZIP to see what works and what does not makes sense too.  We can try, adapt, change, and try again.  

I plan to vote YES to study more.

**Reply by Shazzamazzash** [reply]
Thank you George, I fully support this initiative as it's in the best interest of the community to strive for a ZIP process that aligns with our core values. 

It seems that at least part of the active community agrees that a 1/6 pillar vote does not constitute a sufficient consensus quorum for a ZIP: https://twitter.com/cryptocheshire/status/1564487369335754758?t=YffTi1rUFoGLunk9wjOobA&s=19

**Reply by sultanofstaking** [reply]
100% agree with this "Ultimately, I think it is likely that the way we settle on a de facto framework is by actually trying to push a piece of work through and seeing what works and what the community, pillars, builders, users, will opt-in to.

It is not strictly necessary to start with a ZIP to define the ZIP process. Rather it can first be worked out in practice and then formalized later."

Many pillars will not lean in until this becomes lucrative enough for them to prioritize their time, but we can't wait for them.

**Comment by georgezgeorgez** [comment]
So it's been a while since I've touched this thread.
Our vote did go through and we will continue to share our ideas around ZIPs.

@0x3639 and @Shazzamazzash have actually taken it upon themselves to formalize some of the ideas in this discussion into [ZIP:deeZNNutz-0001](https://github.com/deeZNNutz-com/ZIP.deeZNNutz/blob/main/0001.md). 

We will need to see if it will be adopted by other pillars in the community. 

There has also been some technical work defined to decentralize on-chain spork creation and activation that I will discuss later. 

So far in this thread I've mainly discussed just the Storing and sharing of ideas. During our discussions for ZNN:deeZNNutz-0001, we got deeper into aspects of Acceptance and Activation and those are reflected in the proposal. 

I want to approach the ZIP discussion systematically and continue to discuss the other areas.

2. Feedback and discussion
- I don't think there's really too much guidance we need to give here. Commentary will happen across many channels no matter what
- I do think we should suggest that all ZIPs point to where the ZIP creators will post all "official" updates or where questions can be asked. It seems like EIPs in Ethereum have a "Discussions-To" field that does this. 
- This could be a thread on a forum, a chat room, just to provide a starting point for someone looking to discuss the matter

**Reply by 0x3639** [reply]
Good idea on Feedback / Discussion.  Here is an example of the "Discussions-To":  https://github.com/djrtwo/EIPs/issues/5

**Reply by trevorleazy** [reply]
Looks like a good way to do it. Not like its a complicated step in the ZIP process but does seem like an important one for some community involvement, maybe we could try to acknowledge this in the community roadmap making it a goal to encourage and grow the conversation where possible, organize some informal talk on spaces or in chat then submit anything relevant to the forum.
